### PR TITLE
feat(app): allow drag&drop file upload in browser

### LINKED ISSUE #414: allow browser upload drag & drop
- [x] drag&drop image on entires
  - [x]  add test for entries.ts
- [x] drag&drop image on upload page
- [x] prevent default browser action when dropping somewhere else
- [x] accept only images
- [x] **timeboxed 2h**: styling of drop target
  - [x] on entries page
  - [x]  on upload page
- [x] **timeboxed 1h**: change mouse pointer to copy instead of move
- [x] e2e test
  - [x] drop event test
  - [x] drag event (change style) test

### PR DESCRIPTION
**Pull request by nikaluus**
This feature helps users to easily upload files in browser via drag&drop. It is convenient to upload images from SD-cards and desktop. Files can be dragged into an entry in the entriespage or into the uploadpage.

Closes #414

### CODE REVIEW

File: src/providers/browser-fileupload-selector-service.spec.ts [hunk 1]
```diff
@@ -14,21 +15,59 @@ describe('Provider: BrowserFileuploadSelectorService', () => {
...
+
+  it('should throw ims file type error when file type is not allowed', inject([BrowserFileuploadSelectorService], (browserFileuploadSelectorService: BrowserFileuploadSelectorService) => {
+    let fileList: FileList = Object.setPrototypeOf([], FileList);
+    let fileName = 'invalid.txt';
+    let file: File = new File([new Blob()], fileName, {type: 'text/plain'});
+    fileList[0] = file;
+    expect(() => browserFileuploadSelectorService.getImageFromFileList(fileList)).toThrowError(ImsFileTypeError);
+  }));
+
+  it('should return an image when a event from file dropper called', inject([BrowserFileuploadSelectorService], (browserFileuploadSelectorService: BrowserFileuploadSelectorService) => {
```
Comment (nikaluus) [inline src/providers/browser-fileupload-selector-service.spec.ts]: typo: should be "an" event

File: src/providers/browser-fileupload-selector-service.spec.ts [hunk 2]
```diff
@@ -14,21 +15,59 @@ describe('Provider: BrowserFileuploadSelectorService', () => {
...
+    let file1: File = new File([new Blob()], fileName1, {type: browserFileuploadSelectorService.allowedFileTypes[0]});
+    let file2: File = new File([new Blob()], fileName2, {type: browserFileuploadSelectorService.allowedFileTypes[0]});
+    let event = { dataTransfer: { files: [file1, file2] } };
+    let referenceImage = new Image(fileName1, fileURI, file1);
+    spyOn(window.URL, 'createObjectURL').and.returnValue(fileURI);
+    let returnedImage = browserFileuploadSelectorService.getImageFromFileDrop(event);
+    expect(returnedImage).toEqual(referenceImage);
+  }));
+
+  it('should return an undefined object when a event from file picker with no files is called', inject([BrowserFileuploadSelectorService], (browserFileuploadSelectorService: BrowserFileuploadSelectorService) => {
```
Comment (nikaluus) [inline src/providers/browser-fileupload-selector-service.spec.ts]: typo: should be "an" event

File: src/providers/drag-event-service.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+  public preventEventsOnBody(renderer: Renderer2) {
+    renderer.listen('body', 'dragenter', event => this.preventDefaultDragAction(event));
+    renderer.listen('body', 'dragover', event => this.preventDefaultDragAction(event));
+    renderer.listen('body', 'dragleave', event => this.preventDefaultDragAction(event));
+    renderer.listen('body', 'dragend', event => this.preventDefaultDragAction(event));
+    renderer.listen('body', 'drop', event => this.preventDefaultDragAction(event));
+  }
+
+  public handleDragEvent(event: DragEvent, enterFunction: Function, leaveFunction: Function) {
```
Comment (nikaluus) [inline src/providers/drag-event-service.ts]: handle drop event within this function

File: src/mocks/drag-event-creator.spec.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+export class DragEventCreator {
```
Comment (nikaluus) [inline src/mocks/drag-event-creator.spec.ts]: Rework this helperclass (rename, replace, combine data transfer property, try to remove spy)

File: src/pages/entries/entries.scss [hunk 5]
```diff
@@ -1,3 +1,10 @@
 page-entries {
+  .drag {
```
Comment (nikaluus) [inline src/pages/entries/entries.scss]: rename to drag-overlay (like in uploadpage), adjust also e2e test

File: src/pages/entries/entries.spec.ts [hunk 6]
```diff
@@ -289,4 +291,51 @@ describe('Page: Entries', () => {
...
+    let entryTitle = 'title';
+    spyOn(browserFileuploadSelectorService, 'getImageFromFileDrop').and.returnValue(droppedImage);
+    spyOn(page, 'pushToUploadPageWithPicture').and.callThrough();
+    page.receiveDrop(event, parentImageEntryId, entryTitle);
+    expect(page.pushToUploadPageWithPicture).toHaveBeenCalledWith(droppedImage, parentImageEntryId, entryTitle);
+  }));
+
+
+  it('Should not go to uploadpage after receiving no dropped image', inject([BrowserFileuploadSelectorService], (browserFileuploadSelectorService: BrowserFileuploadSelectorService) => {
+    let event: DragEvent = new DragEventMock('drop', document.createElement('div'));
```
Comment (mleu) [inline src/pages/entries/entries.spec.ts]: second argument to DragEventMock may not be needed

**Pull request merged**